bundles/org.openhab.binding.solarmax/src/main/java/org/openhab/binding/solarmax/internal/connector/SolarMaxConnector.java

```diff
@@ -0,0 +1,396 @@
+        // get the data from the SolarMax device. If we didn't get as many values back as we asked for, there were
+        // communications problems, so set communicationSuccessful appropriately
+
+        Map<SolarMaxCommandKey, String> valuesFromSolarMax = getValuesFromSolarMax(host, port, commandList);
+        boolean allCommandsAnswered = true;
+        for (SolarMaxCommandKey solarMaxCommandKey : commandList) {
+            if (!valuesFromSolarMax.containsKey(solarMaxCommandKey)) {
+                allCommandsAnswered = false;
+                break;
+            }
+        }
+        solarMaxData.setDataDateTime(ZonedDateTime.now());
+        solarMaxData.setCommunicationSuccessful(allCommandsAnswered);
+        solarMaxData.setData(valuesFromSolarMax);
+
+        return solarMaxData;
+    }
+
+    /**
+     * gets values from the SolarMax device addressable at host:port
+     * 
+     * @param host hostname or ip address of the SolarMax device to be contacted
+     * @param port port the SolarMax is listening on (default is 12345)
+     * @param commandList a list of commands to be sent to the SolarMax device
+     * @return
+     * @throws UnknownHostException if the host is unknown
+     * @throws SolarMaxException if some other exception occurs
+     */
+    private static Map<SolarMaxCommandKey, String> getValuesFromSolarMax(final String host, int port,
+            final List<SolarMaxCommandKey> commandList) throws SolarMaxException {
+
+        Socket socket;
+
+        Map<SolarMaxCommandKey, String> returnMap = new HashMap<>();
+
+        // SolarMax can't answer correclty if too many commands are send in a single request, so limit it to 16 at a
+        // time
+        int maxConcurrentCommands = 16;
+        int requestsRequired = (commandList.size() / maxConcurrentCommands);
+        if (commandList.size() % maxConcurrentCommands != 0) {
+            requestsRequired = requestsRequired + 1;
+        }
+        for (int requestNumber = 0; requestNumber < requestsRequired; requestNumber++) {
+            logger.debug("    Requesting data from {}:{} with timeout of {}ms", host, port, responseTimeout);
+
+            int firstCommandNumber = requestNumber * maxConcurrentCommands;
+            int lastCommandNumber = (requestNumber + 1) * maxConcurrentCommands;
+            if (lastCommandNumber > commandList.size()) {
+                lastCommandNumber = commandList.size();
+            }
+            List<SolarMaxCommandKey> commandsToSend = commandList.subList(firstCommandNumber, lastCommandNumber);
+
+            try {
+                socket = getSocketConnection(host, port);
+            } catch (UnknownHostException e) {
+                throw new SolarMaxConnectionException(e);
+            }
+            returnMap.putAll(getValuesFromSolarMax(socket, commandsToSend));
+
+            // SolarMax can't deal with requests too close to one another, so just wait a moment
+            try {
+                Thread.sleep(10);
```

Comment:
Small values can be skipped by the scheduler entirely. To make this work it should be at least 100ms. However, this will block a shared thread in the framework. You could schedule a task to do the next request.

Reply:
This code is proven to work. Waiting 100ms instead of 10ms will significantly increase the delay between the values being read, meaning they're less likely to actually make sense, because they were read at significantly different times. No action taken.

Reply:
I understand your point, that these values should be read at the same time at best. But a delay of less than 100ms can be skipped (i.e. results in no sleep), if the machine has high CPU, IO or interrupt load. However, I won't insist on changing this, because this is not a functional review, but you're warned.

Reply:
Thanks for the warning. A quick search only showed me a bunch of old bugs in this space, but no indication that this was expected behaviour, nor that the bugs are still present. If you have a further reference for me (particularly one with a nice solution that doesn't just waste CPU power), I could look at implementing it.

Reply:
There's no chance to implement a reliable delay shorter than 100ms but longer than 0ms, because the JVM has no real-time capabilities. You can either use 100ms, which slows things down or use a smaller value, which might get skipped, sometimes.